drivers/usb/udc/udc_dwc2.c

```diff
@@ -46,11 +46,13 @@ K_MSGQ_DEFINE(drv_msgq, sizeof(struct dwc2_drv_event),
 	      CONFIG_UDC_DWC2_MAX_QMESSAGES, sizeof(void *));
 
 
-/* Minimum RX FIFO size in 32-bit words considering the largest used OUT packet
- * of 512 bytes. The value must be adjusted according to the number of OUT
- * endpoints.
+/* Default Rx FIFO size in 32-bit words calculated to support:
+ *   * 1 control endpoint in Completer/Buffer DMA mode: 13 locations
+ *   * Global OUT NAK: 1 location
+ *   * Space for 3 * 1024 packets: ((1024/4) + 1) * 3 = 774 locations
+ * Driver adds 2 locations for each OUT endpoint to this value.
  */
-#define UDC_DWC2_GRXFSIZ_DEFAULT	(15U + 512U/4U)
+#define UDC_DWC2_GRXFSIZ_DEFAULT	(13 + 1 + 774)
```

Comment:
This will take up all available space on full-speed devices and they also do not support high bandwidth transfers, e.g. full-speed-only controllers on STM32F413 have only 2048 bytes of FIFO. I thought about different definitions for full/high speed and some Kconfig option, but it would be quite difficult to handle and not good to maintain. I think it is better to adjust it at runtime based on the configuration, e.g. in the udc_enable() implementation.

Reply:
For a proper fifo sizing, the driver must be aware of configurations and do allocations based on that. Because the RxFIFO size must be configured (thus must be known) before the configuration is selected, driver must have prior knowledge of the configurations to know what can be afforded. The configurations would also be necessary to determine whether the thresholding is necessary or not (it is recommended to not enable thresholding unless necessary).

For the time being, to not break STM32F413, as a temporary measure I can make it "dynamic" in `udc_dwc2_init_controller()`. The "dynamic" is because it would just be based on PHY speed and not based on the actual configuration (which is not exposed to UDC drivers).